k2/python/k2/rnnt_loss.py

```diff
@@ -0,0 +1,684 @@
+        (B, S+1, C), i.e. batch, symbol_seq_len+1, num_classes
+     am: acoustic-model part of unnormalized log-probs of symbols, with shape
+       (B, T, C), i.e. batch, frame, num_classes
+     symbols: the symbol sequences, a LongTensor of shape [B][S], and elements in {0..C-1}.
+     termination_symbol: the termination symbol, with 0 <= termination_symbol < C
+     boundary: a LongTensor of shape [B, 4] with elements interpreted as
+        [begin_symbol, begin_frame, end_symbol, end_frame] that is treated as [0, 0, S, T]
+        if boundary is not supplied.
+        Most likely you will want begin_symbol and begin_frame to be zero.
+   Returns:
+      a Tensor of shape (B,), containing the NEGATED total RNN-T loss values
+      for each element of the batch (like log-probs of sequences).
+    """
+    px, py = get_rnnt_logprobs(lm, am, symbols, termination_symbol)
+    return mutual_information_recursion(px, py, boundary)
+
+
+def get_rnnt_logprobs_joint(joint: Tensor, symbols: Tensor,
+                            termination_symbol: int) -> Tuple[Tensor, Tensor]:
+    """
+    Reduces RNN-T problem to a compact, standard form that can then be given
+    (with boundaries) to mutual_information_recursion().  This function is called from
+    rnnt_loss().
+
+    Args:
+      joint: The output of joiner network, with shape
+             (B, T, S + 1, C), i.e. batch, time_seq_len, symbol_seq_len+1, num_classes
+      symbols: A LongTensor of shape [B][S], containing the symbols at each position
+          of the sequence, possibly including EOS
+      termination_symbol: The identity of the termination symbol, must be
+           in {0..C-1}
+    Returns: (px, py) (the names are quite arbitrary).
+              px: logprobs, of shape [B][S][T+1]
+              py: logprobs, of shape [B][S+1][T]
+          in the recursion:
+             p[b,0,0] = 0.0
+             p[b,s,t] = log_add(p[b,s-1,t] + px[b,s-1,t],
+                                p[b,s,t-1] + py[b,s,t-1])
+          .. where p[b][s][t] is the "joint score" of the pair of subsequences of
+          length s and t respectively.  px[b][s][t] represents the probability of
+          extending the subsequences of length (s,t) by one in the s direction,
+          given the particular symbol, and py[b][s][t] represents the probability
+          of extending the subsequences of length (s,t) by one in the t direction,
+          i.e. of emitting the termination/next-frame symbol.
+
+          px[:,:,T] equals -infinity, meaning on the "one-past-the-last" frame
+          we cannot emit any symbols.  This is simply a way of incorporating
+          the probability of the termination symbol on the last frame.
+    """
+    assert joint.ndim == 4
+    (B, T, S1, C) = joint.shape
+    S = S1 - 1
+    assert symbols.shape == (B, S)
+
+    max_value = torch.max(joint)
+    normalizers = (joint - max_value)
+    normalizers = torch.logsumexp(normalizers, dim=3)
```

Comment:
`torch.logsumexp()` is numerically stable, so there is no need to subtract the `max` here.

See https://pytorch.org/docs/stable/generated/torch.logsumexp.html
> Returns the log of summed exponentials of each row of the input tensor in the given dimension dim. The computation is numerically stabilized.

and see its internal implementation
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L1260

```cpp
static Tensor& logsumexp_out_impl(Tensor& result, const Tensor& self, IntArrayRef dims, bool keepdim) {
  // can't take max of empty tensor
  if (self.numel() != 0) {
    auto maxes = at::amax(self, dims, true);
    auto maxes_squeezed = (keepdim ? maxes : squeeze_multiple(maxes, dims));
    maxes_squeezed.masked_fill_(maxes_squeezed.abs() == INFINITY, 0);
    at::sum_out(result, (self - maxes).exp_(), dims, keepdim);
    result.log_().add_(maxes_squeezed);
  } else {
```

Reply:
Yes, Dan had a comment on this line before, I forgot to delete it.  Will fix the comments you gave these days.  Thanks!

Reply:
I see. It has been fixed in the master.